src/Rules/CheckJobDispatchArgumentTypesCompatibleWithClassConstructorRule.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+declare(strict_types=1);
+
+namespace NunoMaduro\Larastan\Rules;
+
+use Illuminate\Foundation\Bus\Dispatchable;
+use PhpParser\BuilderFactory;
+use PhpParser\Node;
+use PhpParser\Node\Expr\StaticCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\FunctionCallParametersCheck;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPStan\Type\BooleanType;
+
+/** @implements Rule<StaticCall> */
+class CheckJobDispatchArgumentTypesCompatibleWithClassConstructorRule implements Rule
+{
+    /** @var ReflectionProvider */
+    private $reflectionProvider;
+
+    /** @var FunctionCallParametersCheck */
+    private $check;
+
+    public function __construct(
+        ReflectionProvider $reflectionProvider,
+        FunctionCallParametersCheck $check
+    ) {
+        $this->reflectionProvider = $reflectionProvider;
+        $this->check = $check;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getNodeType(): string
+    {
+        return StaticCall::class;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function processNode(Node $node, Scope $scope): array
+    {
+        if (! $node->name instanceof Node\Identifier) {
+            return [];
+        }
+
+        $methodName = $node->name->name;
+
+        if (! in_array($methodName, [
+            'dispatch',
+            'dispatchIf',
+            'dispatchUnless',
+            'dispatchSync',
+            'dispatchNow',
+            'dispatchAfterResponse',
+        ],
+            true)
+        ) {
+            return [];
+        }
+
+        if (! $node->class instanceof Node\Name) {
+            return [];
+        }
+
+        $jobClassReflection = $this->reflectionProvider->getClass($scope->resolveName($node->class));
+
+        if (! $jobClassReflection->hasTraitUse(Dispatchable::class)) {
+            return [];
+        }
+
+        if (! $jobClassReflection->hasConstructor()) {
+            $requiredArgCount = 0;
+
+            if (in_array($methodName, ['dispatchIf', 'dispatchUnless'], true)) {
+                $requiredArgCount = 1;
+            }
+
+            if (count($node->getArgs()) > $requiredArgCount) {
+                return [
+                    RuleErrorBuilder::message(sprintf(
+                        'Job class %s does not have a constructor and must be dispatched without any parameters.',
+                        $jobClassReflection->getDisplayName()
+                    ))->build(),
+                ];
+            }
+
+            return [];
+        }
+
+        if (in_array($methodName, ['dispatchIf', 'dispatchUnless'], true)) {
+            if (count($node->getArgs()) < 1) {
```

Comment:
```suggestion
            if ($node->getArgs() === []) {
```
Could we be explicit?

Reply:
Sure. Thanks.